scm-plugins/scm-hg-plugin/src/main/java/sonia/scm/repository/hooks/DefaultHookHandler.java

```diff
@@ -0,0 +1,156 @@
+import sonia.scm.repository.spi.HgHookContextProvider;
+import sonia.scm.repository.spi.HookEventFacade;
+import sonia.scm.security.BearerToken;
+import sonia.scm.security.CipherUtil;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.Socket;
+import java.util.List;
+
+import static java.util.Collections.singletonList;
+
+class DefaultHookHandler implements HookHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultHookHandler.class);
+
+  private final HookEventFacade hookEventFacade;
+  private final HookEnvironment environment;
+  private final HookContextProviderFactory hookContextProviderFactory;
+  private final Socket socket;
+
+  @Inject
+  public DefaultHookHandler(HookContextProviderFactory hookContextProviderFactory, HookEventFacade hookEventFacade, HookEnvironment environment, @Assisted Socket socket) {
+    this.hookContextProviderFactory = hookContextProviderFactory;
+    this.hookEventFacade = hookEventFacade;
+    this.environment = environment;
+    this.socket = socket;
+  }
+
+  @Override
+  public void run() {
+    LOG.trace("start handling hook protocol");
+    try (InputStream input = socket.getInputStream(); OutputStream output = socket.getOutputStream()) {
+      handleHookRequest(input, output);
+    } catch (IOException e) {
+      LOG.warn("failed to read hook request", e);
+    } finally {
+      LOG.trace("close client socket");
+      close();
+    }
+  }
+
+  private void handleHookRequest(InputStream input, OutputStream output) throws IOException {
+    Request request = Sockets.read(input, Request.class);
+    Response response = handleHookRequest(request);
+    Sockets.send(output, response);
+  }
+
+  private Response handleHookRequest(Request request) {
+    LOG.trace("process {} hook for node {}", request.getType(), request.getNode());
+    try {
+      if (!environment.isAcceptAble(request.getChallenge())) {
+        LOG.warn("received hook with invalid challenge: {}", request.getChallenge());
+        return error("invalid hook challenge");
+      }
+
+      authenticate(request);
+      environment.setPending(request.getType() == RepositoryHookType.PRE_RECEIVE);
+
+      HgHookContextProvider context = hookContextProviderFactory.create(request.getRepositoryId(), request.getNode());
+      hookEventFacade.handle(request.getRepositoryId()).fireHookEvent(request.getType(), context);
+
+      return new Response(context.getHgMessageProvider().getMessages(), false);
+    } catch (AuthenticationException ex) {
+      LOG.warn("hook authentication failed", ex);
+      return error("hook authentication failed");
+    } catch (NotFoundException ex) {
+      LOG.warn("could not find repository with id {}", request.getRepositoryId(), ex);
+      return error("repository not found");
+    } catch (Exception ex) {
+      LOG.warn("unknown error on hook occurred", ex);
+      return error("unknown error");
```

Comment:
I've applied your suggestion, but i've passed the message of the exception instead of "unkown error". This restores the behaviour before that change, but i'm not sure if this is the best way. Because we could leak technical details on unwanted exceptions.

Reply:
The leak is exactly why i've left it the way you've done it before for unknown exceptions. If you want to have proper messages, one should use `ExceptioWithContext`.